tests/unit/test_charm.py

```diff
@@ -344,13 +345,86 @@ def bundle_handler(args: ops.testing.ExecArgs) -> None:
     charm: DiscourseCharm = typing.cast(DiscourseCharm, harness.charm)
 
     email = "sample@email.com"
-    password = "somepassword"  # nosec
     event = MagicMock(spec=ActionEvent)
-    event.params = {
-        "email": email,
-        "password": password,
-    }
-    charm._on_add_admin_user_action(event)
+    event.params = {"email": email}
+    charm._on_promote_user_action(event)
+    assert expected_exec_call_was_made
+
+
+@patch.object(ops.Container, "exec")
```

Comment:
the ops library already has `harness.handle_exec`. This function has clear interfaces and is easy to patch in a similar way to a real exec of pebble. There is no need to patch the exec in the container class.

Reply:
Nice to know, I was looking at [indico](https://github.com/canonical/indico-operator/blob/main/tests/unit/test_core.py#L71) for reference. Your suggestion has been implemented!